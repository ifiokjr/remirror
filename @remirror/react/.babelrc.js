const defaultConfig = require('../../support/babel/base.babel');

module.exports = {
  ...defaultConfig,
  presets: [
    [
      '@babel/preset-env',
      {
        modules: false,
        targets: {
          node: '8',
        },
      },
    ],
    '@babel/preset-typescript',
  ],
  plugins: [...defaultConfig.plugins],
};
