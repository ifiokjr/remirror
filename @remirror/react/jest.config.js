const config = require('../../config/jest.config');
const { join } = require('path');

module.exports = {
  ...config,
  name: '@remirror/react',
  displayName: 'react',
  setupFilesAfterEnv: [join(__dirname, 'jest.framework.ts')],
  testEnvironment: 'jsdom',
};
