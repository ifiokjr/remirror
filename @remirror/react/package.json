{
  "name": "@remirror/react",
  "description": "react utilities for remirror",
  "license": "MIT",
  "author": "Ifiok Jr. <ifiokotung@gmail.com>",
  "homepage": "https://github.com/remirror/remirror/tree/master/@remirror/react",
  "repository": "https://github.com/remirror/remirror/tree/master/@remirror/react",
  "version": "0.7.6",
  "main": "lib/index.js",
  "module": "lib/dist/react.esm.js",
  "files": [
    "lib",
    "src"
  ],
  "sideEffects": false,
  "types": "lib/index.d.ts",
  "dependencies": {
    "@babel/runtime": "^7",
    "@remirror/core": "^0.8.0",
    "@remirror/core-extensions": "^0.7.5",
    "@remirror/react-portals": "^0.7.4",
    "@remirror/react-ssr": "^0.7.5",
    "@remirror/react-utils": "^0.7.5",
    "@remirror/ui": "^0.7.5",
    "@types/hoist-non-react-statics": "^3.3.0",
    "@types/prosemirror-commands": "^1.0.1",
    "@types/prosemirror-model": "^1.7.2",
    "@types/prosemirror-state": "^1.2.3",
    "@types/prosemirror-view": "^1.11.2",
    "hoist-non-react-statics": "^3.3.0",
    "prosemirror-commands": "^1.0.8",
    "prosemirror-model": "^1.7.4",
    "prosemirror-state": "^1.2.4",
    "prosemirror-view": "^1.12.0"
  },
  "peerDependencies": {
    "@emotion/core": "^10",
    "@types/react": "^16.8.0",
    "@types/react-dom": "^16.8.0",
    "react": "^16.8.0",
    "react-dom": "^16.8.0"
  },
  "publishConfig": {
    "access": "public"
  },
  "cjs": "lib/dist/react.cjs.js",
  "meta": {
    "sizeLimit": "115 KB"
  }
}
