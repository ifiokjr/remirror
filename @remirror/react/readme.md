<div align="center">
	<br />
	<div align="center">
		<img width="300" src="https://cdn.jsdelivr.net/gh/ifiokjr/remirror/support/assets/logo-icon.svg" alt="remirror" />
    <h1 align="center">react</h1>
	</div>
    <br />
    <br />
    <br />
    <br />
</div>

[![npm bundle size (scoped)](https://img.shields.io/bundlephobia/minzip/@remirror/react.svg?style=for-the-badge)](https://bundlephobia.com/result?p=@remirror/react) [![npm](https://img.shields.io/npm/dm/@remirror/react.svg?style=for-the-badge&logo=npm)](https://www.npmjs.com/package/@remirror/react) [![Dependencies (path)](https://img.shields.io/david/ifiokjr/remirror.svg?logo=npm&path=@remirror%2Freact&style=for-the-badge)](https://github.com/ifiokjr/remirror/blob/master/@remirror/react/package.json) [![NPM](https://img.shields.io/npm/l/@remirror/react.svg?style=for-the-badge)](https://github.com/ifiokjr/remirror/blob/master/LICENSE) [![GitHub issues by-label](https://img.shields.io/github/issues/ifiokjr/remirror/@remirror/react.svg?label=Open%20Issues&logo=github&style=for-the-badge)](https://github.com/ifiokjr/remirror/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+label%3A%40remirror%2Freact) [![GitHub pull requests by-label](https://img.shields.io/github/issues-pr/ifiokjr/remirror/@remirror/react.svg?label=Open%20Pull%20Requests&logo=github&style=for-the-badge)](https://github.com/ifiokjr/remirror/pulls?utf8=%E2%9C%93&q=is%3Apr+is%3Aopen+sort%3Aupdated-desc+label%3A%40remirror%2Freact)

The react components for the remirror editor

## Installation

```bash
yarn add @remirror/react
```

## Usage

For in depth usage with proper code example see the [docs](https://docs.remirror.org)
