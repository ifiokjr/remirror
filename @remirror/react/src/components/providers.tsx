import React, { ComponentType, FunctionComponent } from 'react';

import { Cast, isEmptyObject, MakeOptional, Omit } from '@remirror/core';
import hoistNonReactStatics from 'hoist-non-react-statics';
import { RemirrorEditorContext } from '../contexts';
import { defaultProps } from '../helpers';
import { useRemirrorManagerContext } from '../hooks';
import {
  GetPositionerReturn,
  InjectedRemirrorProps,
  RemirrorElementType,
  RemirrorFC,
  RemirrorProps,
  UsePositionerParams,
} from '../types';
import { Remirror } from './remirror';

export type RemirrorProviderProps = MakeOptional<Omit<RemirrorProps, 'children'>, keyof typeof defaultProps>;

/**
 * The RemirrorProvider which injects context into any of it child components.
 *
 * These can either be consumed using React Hooks
 * - `useRemirrorContext`
 * - `usePositioner`
 *
 * Or the higher order components
 * - `withRemirror`
 * - `withPositioner`
 */
export const RemirrorEditor: RemirrorFC<RemirrorProviderProps> = ({ children, ...props }) => {
  return (
    <Remirror {...props}>
      {value => <RemirrorEditorContext.Provider value={value}>{children}</RemirrorEditorContext.Provider>}
    </Remirror>
  );
};

RemirrorEditor.$$remirrorType = RemirrorElementType.EditorProvider;

/**
 * Renders the content while pulling the manager from the context and passing it on to the
 * RemirrorProvider.
 *
 * If no manager exists the child components are not rendered.
 */
export const ManagedRemirrorEditor: RemirrorFC<Omit<RemirrorProviderProps, 'manager'>> = ({
  children,
  ...props
}) => {
  const manager = useRemirrorManagerContext();

  return manager ? (
    <RemirrorEditor {...props} manager={manager}>
      {children}
    </RemirrorEditor>
  ) : null;
};

ManagedRemirrorEditor.$$remirrorType = RemirrorElementType.ManagedEditorProvider;

const checkValidRenderPropParams = (params: InjectedRemirrorProps) => {
  if (isEmptyObject(params)) {
    throw new Error('No props received for the RemirrorProvider component');
  }
  return true;
};

/**
 * A higher order component which passes the RemirrorContext to the component it wraps.
 *
 * @param WrappedComponent
 */
export const withRemirror = <GProps extends InjectedRemirrorProps>(
  WrappedComponent: ComponentType<GProps>,
) => {
  type EnhancedComponentProps = Omit<GProps, keyof InjectedRemirrorProps>;
  const EnhancedComponent: FunctionComponent<EnhancedComponentProps> = props => {
    return (
      <RemirrorEditorContext.Consumer>
        {params => {
          checkValidRenderPropParams(params);
          return <WrappedComponent {...Cast<GProps>({ ...props, ...params })} />;
        }}
      </RemirrorEditorContext.Consumer>
    );
  };

  return hoistNonReactStatics(EnhancedComponent, WrappedComponent);
};

/**
 * A higher order component which passes the positioner props to the component it wraps.
 * This is useful for creating menus which need access to position information
 *
 * @param params
 */
export const withPositioner = <GRefKey extends string = 'ref'>({
  positioner,
  ...rest
}: UsePositionerParams<GRefKey>) => <GProps extends GetPositionerReturn<GRefKey>>(
  WrappedComponent: ComponentType<GProps>,
) => {
  type EnhancedComponentProps = Omit<GProps, keyof GetPositionerReturn<GRefKey>>;
  const EnhancedComponent: FunctionComponent<EnhancedComponentProps> = props => {
    return (
      <RemirrorEditorContext.Consumer>
        {({ getPositionerProps }) => {
          return (
            <WrappedComponent
              {...Cast<GProps>({ ...props, ...getPositionerProps<GRefKey>({ ...positioner, ...rest }) })}
            />
          );
        }}
      </RemirrorEditorContext.Consumer>
    );
  };

  return hoistNonReactStatics(EnhancedComponent, WrappedComponent);
};
