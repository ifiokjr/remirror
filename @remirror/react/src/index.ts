export * from './node-views';
export * from './components';
export * from './positioners';
export * from './hooks';
export * from './hocs';
export * from '@remirror/react-utils';
