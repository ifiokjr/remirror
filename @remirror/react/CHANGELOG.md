# @remirror/react

## 0.11.0

### Minor Changes

- 026d4238: Add a `focus` method to the remirror editor context object. It allows focusing at a provided
  position which can be `start`, `end`, a specific position or a range using the `{from: number; to: number}`
  type signature.

  To use this run

  ```tsx
  import { useRemirrorContext } from '@remirror/react';

  const MyEditor = () => {
    const { focus, getRootProps } = useRemirrorContext();

    useEffect(() => {
      focus('end'); // Autofocus to the end once
    }, []);
  };
  return <div {...getRootProps()} />;
  ```

  Resolves the initial issue raised in #229.

- 69d00c62: Add custom arguments to `autoFocus` props. The same arguments that can added to the `focus()`
  context method can now be passed as a prop.

### Patch Changes

- Updated dependencies [c2237aa0]
  - @remirror/core@0.11.0
  - @remirror/core-extensions@0.11.0
  - @remirror/react-ssr@0.11.0

## 0.7.7

### Patch Changes

- Updated dependencies [0300d01c]
  - @remirror/core@0.9.0
  - @remirror/core-extensions@0.7.6
  - @remirror/react-utils@0.7.6
  - @remirror/ui@0.7.6
  - @remirror/react-ssr@0.7.6

## 0.7.6

### Patch Changes

- Updated dependencies [24f83413]
- Updated dependencies [24f83413]
  - @remirror/core@0.8.0
  - @remirror/core-extensions@0.7.5
  - @remirror/react-utils@0.7.5
  - @remirror/ui@0.7.5
  - @remirror/react-ssr@0.7.5

## 0.7.5

### Patch Changes

- 7380e18f: Update repository url from ifiokjr/remirror to remirror/remirror to reflect new GitHub
  organisation.
- Updated dependencies [10419145]
- Updated dependencies [7380e18f]
  - @remirror/core-extensions@0.7.4
  - @remirror/core@0.7.4
  - @remirror/react-portals@0.7.4
  - @remirror/react-ssr@0.7.4
  - @remirror/react-utils@0.7.4
  - @remirror/ui@0.7.4

## 0.7.4

### Patch Changes

- 416d65da: Better code comment docs around how to apply additional extensions (#186).

## 0.7.3

### Patch Changes

- 5f85c0de: Bump a new version to test out the changeset API.
- Updated dependencies [5f85c0de]
  - @remirror/core@0.7.3
  - @remirror/core-extensions@0.7.3
  - @remirror/react-portals@0.7.3
  - @remirror/react-ssr@0.7.3
  - @remirror/react-utils@0.7.3
  - @remirror/ui@0.7.3
