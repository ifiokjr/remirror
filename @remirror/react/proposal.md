## Proposal

I've found the current way of dealing with extensions as classes quite cumbersome. It places the burden of configuration on the `user` and is not very `React`ish.

The following is a proposal which makes configuration much simpler.

```ts
import { RemirrorManager, RemirrorEditor } from '@remirror/react';
import { Placeholder, Mention } from '@remirror/core-extensions';
import { MenuComponent, SuggestionsComponent } from './components';

const Editor = () => {
  return (
    <RemirrorManager>
      <Placeholder priority={1} emptyNodeClass='no-placeholder' />
      <Mention priority={1} {...props} />

      <RemirrorEditor>
        <MenuComponent />
        <SuggestionsComponent />
      </RemirrorEditor>
    </RemirrorManager>
  );
};
```

- `RemirrorManager` injects an ExtensionManager instance into the context.
- `RemirrorEditor` inject the `InjectedRenderProps` into the context so that subcomponents can either use hooks and HOC's

The main innovation is enabling extensions to become part of the react UI and automatically creating a manager for any nested RemirrorEditors. This means we can hide implementation details from the user allows the view layer (React) to be responsible for creating the necessary pieces for instantiating a Prosemirror editor.

This also means that we'll need to move away from the default renderProp Provider components are now front and center and the default export to use.

## The current status

At the moment creating a remirror editor is done by adding an extensions props to the remirror editor with all the extensions instantiated with their options.

An example is the current UITwitter component which doesn't look very react-like with.

```ts
<RemirrorEditor extensions={[new Mention({ name: 'mentionHash', ...otherOptions }), new Mention({ name: 'mentionAt', ...otherOptions )]} />
```

## Ideal API

Automated typescript inference is an important goal of this library.

Ideally, the API would have been something along the lines of

- `ExtensionComponent` - A component which takes the Extension constructor as a prop and automatically infers the other props for the React component. It would also

For example:

```tsx
<ExtensionComponent extension={Placeholder} priority={1} emptyNodeClass='no-placeholder' />
```

Unfortunately, while the dynamically created inference works for ordinary class and objects it seems to be lost inside the JSX component and I couldn't get the Typescript types to work automatically.

Because this wasn't possible (or I haven't yet figured out how to make it possible), I've had to break a few of the principles I had wanted to keep while building this project.

- `@remirror/core-extensions` would require a peer dependency on React.
- Each extension now needs to implement a static property `Component`

This all means that the underlying core libraries are coupled to the React view layer.

As soon as possible I'd like to remove this so that extensions can be framework agnostic and the view layers implement components that manage the injection of the mangers into the `RemirrorEditor`.
