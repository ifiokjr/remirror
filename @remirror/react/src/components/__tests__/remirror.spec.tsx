import { fireEvent, render } from '@testing-library/react';
import { axe } from 'jest-axe';
import React from 'react';
import { renderToString } from 'react-dom/server';

import { EditorState, fromHTML } from '@remirror/core';
import { InjectedRemirrorProps, RemirrorEditorStateListenerParams } from '@remirror/react-utils';
import { createTestManager } from '@test-fixtures/schema-helpers';
import { Remirror } from '../';
import { RemirrorEditorProps } from '../providers';

const textContent = `This is editor text`;
const label = 'Remirror editor';
const handlers = {
  onChange: jest.fn(),
  onBlur: jest.fn(),
  onFocus: jest.fn(),
  onFirstRender: jest.fn(),
};

test('should be called via a render prop', () => {
  const mock = jest.fn(() => <div />);
  const { getByLabelText } = render(
    <Remirror manager={createTestManager()} label={label} {...handlers}>
      {mock}
    </Remirror>,
  );
  expect(mock).toHaveBeenCalledWith(expect.any(Object));
  expect(handlers.onFirstRender).toHaveBeenCalledWith(expect.any(Object));
  expect(handlers.onFirstRender.mock.calls[0][0].getText()).toBe('');
  expect(handlers.onFirstRender.mock.calls[0][0].getJSON().doc.type).toBe('doc');
  expect(handlers.onFirstRender.mock.calls[0][0].getHTML().type).toBe(undefined);
  const editorNode = getByLabelText(label);
  expect(editorNode).toHaveAttribute('role', 'textbox');
});

test('TextEditor is accessible', async () => {
  const results = await axe(
    renderToString(<Remirror manager={createTestManager()}>{() => <div />}</Remirror>),
  );
  expect(results).toHaveNoViolations();
});

test('should not render the editor when no render prop available', () => {
  const spy = jest.spyOn(console, 'error').mockImplementation(() => undefined);
  expect(() =>
    // @ts-ignore
    render(<Remirror label={label} manager={createTestManager()} />),
  ).toThrowErrorMatchingInlineSnapshot(`"The child argument to the Remirror component must be a function."`);
  expect(spy).toHaveBeenCalledTimes(2);
  spy.mockRestore();
});

test('should allow text input and fire all handlers', () => {
  let setContent: InjectedRemirrorProps['setContent'] = jest.fn();
  const mock = jest.fn((value: InjectedRemirrorProps) => {
    setContent = value.setContent;
    return <div />;
  });
  const { getByLabelText } = render(
    <Remirror label={label} {...handlers} manager={createTestManager()} stringHandler={fromHTML}>
      {mock}
    </Remirror>,
  );

  setContent(`<p>${textContent}</p>`, true);
  const editorNode = getByLabelText(label);
  expect(handlers.onChange).toHaveBeenCalledTimes(1);
  expect(handlers.onFirstRender.mock.calls[0][0].getText()).toBe(textContent);
  fireEvent.blur(editorNode);
  expect(handlers.onBlur).toHaveBeenCalledTimes(1);
  fireEvent.focus(editorNode);
  expect(handlers.onFocus).toHaveBeenCalledTimes(1);
});

test('can update contenteditable with props', () => {
  const mock = jest.fn(() => <div />);
  const manager = createTestManager();
  const El = ({ editable }: { editable: boolean }) => (
    <Remirror editable={editable} label={label} manager={manager}>
      {mock}
    </Remirror>
  );
  const { rerender, getByLabelText } = render(<El editable={true} />);
  expect(getByLabelText(label)).toHaveAttribute('contenteditable', 'true');
  rerender(<El editable={false} />);
  expect(getByLabelText(label)).toHaveAttribute('contenteditable', 'false');
});

test('should render a unique class on the root document', () => {
  const mock = jest.fn(() => <div />);
  const { getByLabelText } = render(
    <Remirror label={label} {...handlers} manager={createTestManager()}>
      {mock}
    </Remirror>,
  );
  const editorNode = getByLabelText(label);
  expect(editorNode.className).toMatch(/remirror-[A-Za-z0-9_-]{5}/);
});

describe('initialContent', () => {
  it('should render with string content', () => {
    const { container } = render(
      <Remirror
        label={label}
        {...handlers}
        manager={createTestManager()}
        initialContent={'<p>Hello</p>'}
        stringHandler={fromHTML}
      >
        {() => <div />}
      </Remirror>,
    );
    expect(container.innerHTML).toInclude('Hello');
  });

  it('renders with json', () => {
    const content = {
      type: 'doc',
      content: [
        {
          type: 'paragraph',
          content: [
            {
              type: 'text',
              text: 'Hello',
            },
          ],
        },
      ],
    };

    const { container } = render(
      <Remirror label={label} {...handlers} manager={createTestManager()} initialContent={content}>
        {() => <div />}
      </Remirror>,
    );
    expect(container.innerHTML).toInclude('Hello');
  });
});

describe('Remirror Controlled Component', () => {
  const initialContent = `<p>Hello</p>`;
  const expectedContent = `<p>World</p>`;
  let props: RemirrorEditorProps;
  beforeEach(() => {
    props = {
      label,
      manager: createTestManager(),
      initialContent,
      stringHandler: fromHTML,
    };
  });

  it('should call onStateChange', () => {
    let value: EditorState | null = null;
    const onStateChange = jest.fn<void, [RemirrorEditorStateListenerParams]>(params => {
      value = params.newState;
    });
    render(
      <Remirror {...props} value={value} onStateChange={onStateChange}>
        {() => <div />}
      </Remirror>,
    );

    expect(onStateChange).toHaveBeenCalled();
    expect(value).not.toBeNull();
    expect(value!.doc.toJSON()).toMatchSnapshot();
  });

  it('should only update the state when value changes', () => {
    let testValue: EditorState | null = null;
    let setContent: InjectedRemirrorProps['setContent'] = jest.fn();
    const onStateChange = jest.fn<void, [RemirrorEditorStateListenerParams]>(params => {
      testValue = params.newState;
    });

    const mock = jest.fn((params: InjectedRemirrorProps) => {
      setContent = params.setContent;
      return <div />;
    });

    const Cmp = ({ value }: { value?: EditorState | null }) => (
      <Remirror {...props} onStateChange={onStateChange} value={value}>
        {mock}
      </Remirror>
    );
    const { rerender, getByRole } = render(<Cmp value={null} />);

    setContent(expectedContent, true);
    rerender(<Cmp value={null} />);
    expect(getByRole('textbox')).toContainHTML(initialContent);

    setContent(expectedContent, true);
    rerender(<Cmp value={testValue} />);
    expect(getByRole('textbox')).toContainHTML(expectedContent);
  });
});
