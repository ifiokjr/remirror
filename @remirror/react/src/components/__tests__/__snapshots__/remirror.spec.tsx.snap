// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Remirror Controlled Component should call onStateChange 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "text": "Hello",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;
